### PR TITLE
Add a lineReader to provide two-line lookahead

### PR DESCRIPTION
**Pull request by mrnugget**

### CODE REVIEW

File: diff/reader_util.go [hunk 1]
```diff
@@ -2,9 +2,77 @@ package diff
...
+	return next, nil
+}
+
+// nextLineStartsWith looks at the line that would be returned by the next call
+// to readLine to check whether it has the given prefix.
+//
+// io.EOF and bufio.ErrBufferFull errors are ignored so that the function can
+// be used when at the end of the file.
+func (l *lineReader) nextLineStartsWith(prefix string) (bool, error) {
+	return l.lineHasPrefix(l.cachedNextLine, prefix, l.cachedNextLineErr)
```
Comment (eseliger) [inline diff/reader_util.go:46]: if someone was to use it before using readLine() once, this would fail, right? Maybe that should be added to a comment?

**Review by eseliger — APPROVED** [review]
Cool, makes sense to me to look two lines ahead.

File: diff/reader_util.go (hunk 1 above)
Comment (mrnugget) [reply diff/reader_util.go:46]: Good point! I think it makes sense to advance the buffer when nothing has been read, wdyt?

Comment (eseliger) [reply diff/reader_util.go:46]: Sounds good!

**Review by keegancsmith — APPROVED** [review]
I still feel a bit iffy about error conditions here, but its a clear improvement over previous code. Great thing is we have a great test case that is sourcegraph.com, so in practice this is fine :D

**Comment by mrnugget** [comment]
> Great thing is we have a great test case that is sourcegraph.com, so in practice this is fine :D

That's the spirit!

I want to add more tests to this PR to check for more conditions. I also think it makes sense to add a more integration-level test here that makes sure that a huge diff can be parsed correctly. We have unit tests for the subparts, but I think a test that checks for multi-file diffs, including hunk contents, is missing.

**Pull request merged**